### PR TITLE
Use hard-coded server for roles install

### PR DESCRIPTION
**Pull request by AlanCoding**
##### SUMMARY
Automation Hub does not host roles, only collections, and this command will error if role requirements are present.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION
```
9.0.1
```

### CODE REVIEW

**Comment by jborean93** [comment]
An alternative is in ansible itself is to filter the server list until we reach one that supports the `v1` API that roles require.

**Comment by jborean93** [comment]
Here is a related issue I've opened on ansible/ansible https://github.com/ansible/ansible/issues/65440

**Comment by ryanpetrello** [comment]
@AlanCoding do we still need this PR?

**Comment by AlanCoding** [comment]
@jborean93 is the issue https://github.com/ansible/ansible/issues/65440 addressed by the merge https://github.com/ansible/ansible/pull/67843 in any way?

A continuation on that question, with https://github.com/ansible/awx/pull/7082 we changed the role install to use `ansible-galaxy role install`. Does this buy us anything? Will that command still error if the Galaxy server list is Automation Hub as 1st priority and then Galaxy as 2nd priority?

Is there any other way to configure the settings so that both role and collection installs work with Automation Hub as the primary?

This is probably still a major bug.

**Comment by jborean93** [comment]
> is the issue ansible/ansible#65440 addressed by the merge ansible/ansible#67843 in any way?

No, ansible/ansible#67843 just tries to install both collections and roles in a requirements file when running `ansible-galaxy install -r requirements.yml`. It still follows the same behaviour for selecting a galaxy server (if required) from the config.

> A continuation on that question, with #7082 we changed the role install to use ansible-galaxy role install. Does this buy us anything?

Depends on how you define the server when calling `ansible-galaxy role install`. From a brief look at the PR it seems like this would still be needed, i.e. you should not enable AH servers when installing roles because AH does not provide support for installing roles.

> Is there any other way to configure the settings so that both role and collection installs work with Automation Hub as the primary?

The only way would be to explicitly set what server you want to run with during install time. This can be done using the `--server <galaxy_server_in_config or URL>` option or setting `ANSIBLE_GALAXY_SERVER_LIST=<galaxy_server_in_config>`.

> This is probably still a major bug.

Maybe, adding support for a server fallback scenario on `ansible-galaxy role install` like we do with collections has some complications

* The code for roles is heavily geared towards a requirement being linked to a single server, changing that is a pretty invasive change
* This only affects roles downloaded from a Galaxy instance, if the requirement is a git URL it shouldn't matter. Do we even know how many orgs actually allow access to the public Galaxy server, my guess is that it's not too many
* We don't have very good integration tests for roles, every time we touch the code usually something breaks that we didn't think of at the time, this leads to a very small appetite for touching this code
* The Galaxy replacement galaxy_ng is not planning on [supporting roles](https://github.com/ansible/galaxy_ng/issues/58). This means that this could have a limited lifetime and we are better off allocating resources around making roles in a collection better for users instead of improving the role experience

That's not to say it isn't possible, this was just never made a priority for us.

**Comment by AlanCoding** [comment]
Then it may make the most sense to continue with this PR, and maybe even add a warning that it won't be supported forever.

File: awx/main/conf.py
Comment (ryanpetrello) [inline awx/main/conf.py:448]: `avilable`

**Review by ryanpetrello — APPROVED** [review]

**Comment by wenottingham** [comment]
If someone has managed to deploy a version of old-galaxy pointed at their github enterprise, this would break their setup, correct?

**Comment by ryanpetrello** [comment]
@wenottingham,

> If someone has managed to deploy a version of old-galaxy pointed at their github enterprise, this would break their setup, correct?

...that's a thing?  Yes, this will break that.

**Comment by AlanCoding** [comment]
I heard @chrismeyersfsu had some experience with that, but I thought it never got beyond "hey, maybe this might conceptually work"

**Comment by wenottingham** [comment]
I don't know if that's a thing, to be honest; but it might be.

More concerningly, I'd rather not merge this and then have to undo it if https://github.com/ansible/ansible/issues/65440#issuecomment-651178575 is going to land.

**Comment by ryanpetrello** [comment]
@wenottingham is https://github.com/ansible/ansible/issues/65440#issuecomment-651178575 something that would be back portable to 2.9.x, though?  My impression from speaking with @sivel is "no".

**Comment by wenottingham** [comment]
If 2.9 is intended for LTS for however many months, it should be.

**Comment by AlanCoding** [comment]
> I'd rather not merge this and then have to undo it if ansible/ansible#65440 (comment) is going to land.

Why can't we merge both? I don't see anything in that diff which makes me think this wouldn't work in conjunction with it. The only thing that would undermine this change is what you mentioned before, "old-galaxy pointed at their github enterprise". The server wasn't even customizable in AWX until very recently. If someone was doing that, it would be a wacky combination of versions.

**Comment by sivel** [comment]
I'm fine with backporting the fix I put in 65540.  It's minimally invasive.

**Comment by AlanCoding** [comment]
Excellent, we can revive this if circumstances require.

**Comment by ryanpetrello** [comment]
@sivel I think I'm gonna ask @kdelee or @squidboylan to do some verification on the AWX side to make sure I'm not being bone-headed and have this wrong.  But I'd love for your patch to be what we land and backport here, because I think it's the sanest, simplest approach to fix it.

**Pull request closed without merging**
